### PR TITLE
Parse Integer negative value validation

### PR DESCRIPTION
**Pull request by TrevisGordan**
## Overview

- Expands function parse_integer with an optional negative value check and
expands function parse_integer_from_args to raise INVALID_PARAM on a forbidden negative value.

- Adds limit parameter negative value validation check
to fix https://github.com/element-hq/synapse/issues/16918 (500 internal server error on negative limit parameter)

- Adds negative value validation to parse_integer functions.
Removing duplicate negative value check logics blocks.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Review by reivilibre — APPROVED** [review]
Thanks!

File: changelog.d/16920.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Adds `limit` parameter negative value validation check to prevent 500 internal server error on publicRooms request.
```
Comment (reivilibre) [inline changelog.d/16920.bugfix:1]:
```suggestion
Adds validation to ensure that the `limit` parameter on `/publicRooms` is non-negative.
```

File: synapse/http/servlet.py [hunk 2]
```diff
@@ -70,12 +70,21 @@ def parse_integer(request: Request, name: str, *, required: Literal[True]) -> in
...
 def parse_integer(
-    request: Request, name: str, default: Optional[int] = None, required: bool = False
-) -> Optional[int]: ...
+    request: Request,
+    name: str,
+    default: Optional[int] = None,
+    required: bool = False,
+    negative: bool = False,
+) -> int:
+    ...
```
Comment (dklimpel) [inline synapse/http/servlet.py:79]:
```suggestion
) -> int: ...
```

File: tests/rest/admin/test_media.py [hunk 3]
```diff
@@ -277,7 +277,7 @@ def test_missing_parameter(self) -> None:
         self.assertEqual(400, channel.code, msg=channel.json_body)
         self.assertEqual(Codes.MISSING_PARAM, channel.json_body["errcode"])
         self.assertEqual(
-            "Missing integer query parameter 'before_ts'", channel.json_body["error"]
+            "Missing required integer query parameter before_ts", channel.json_body["error"]
```
Comment (dklimpel) [inline tests/rest/admin/test_media.py:280]:
```suggestion
            "Missing required integer query parameter before_ts",
            channel.json_body["error"],
```

File: synapse/rest/admin/media.py [hunk 4]
```diff
@@ -311,29 +311,18 @@ async def on_POST(
...
                 + "Double check that you are providing a timestamp in milliseconds.",
                 errcode=Codes.INVALID_PARAM,
             )
-        if size_gt < 0:
-            raise SynapseError(
-                HTTPStatus.BAD_REQUEST,
-                "Query parameter size_gt must be a string representing a positive integer.",
-                errcode=Codes.INVALID_PARAM,
-            )
+
```
Comment (dklimpel) [inline synapse/rest/admin/media.py:325]:
```suggestion
```

**Review by dklimpel — COMMENTED** [review]
some hints :)

File: synapse/http/servlet.py [hunk 5]
```diff
@@ -70,12 +70,20 @@ def parse_integer(request: Request, name: str, *, required: Literal[True]) -> in
...
 @overload
 def parse_integer(
-    request: Request, name: str, default: Optional[int] = None, required: bool = False
-) -> Optional[int]: ...
+    request: Request,
+    name: str,
+    default: Optional[int] = None,
+    required: bool = False,
+    negative: bool = False,
+) -> int: ...
```
Comment (dklimpel) [inline synapse/http/servlet.py:78]:
```suggestion
) -> Optional[int]: ...
```
Or why this was changed?

Comment (TrevisGordan) [reply synapse/http/servlet.py]: Originally it was changed to satisfy mypy.
But you are right; there are cases where these would not be correct.
I added (fixed) those overloads to satisfy those signatures, and make mypy happy.
Thanks for the hint!

**Review by reivilibre — APPROVED** [review]
Thanks for fixing this up!

**Comment by TrevisGordan** [comment]
Hey @reivilibre - I believe the prior systest fails were caused by outdated upstream mismatch.
After re-syncing / merging changes, the systests do pass 😳.
Could you just restart the test again - please (...before this is outdated again 🙈 Thank you! )

**Pull request merged**